sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -588,6 +588,16 @@ object ColumnPruning extends Rule[LogicalPlan] {
         .map(_._2)
       p.copy(child = g.copy(child = newChild, unrequiredChildIndex = unrequiredIndices))
 
+    // prune unrequired nested fields
+    case p @ Project(projectList, g: Generate) =>
```

Comment:
Why we need to special case `Generate`? I see there is a general case for nested column pruning below.

Reply:
I think this is more general case you are looking for:

https://github.com/apache/spark/blob/163a6e298213f216f74f4764e241ee6298ea30b6/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala#L623-L624

In the general case, it's doing pruning only when the flag (nestedSchemaPruningEnabled) is enabled. The case considers some operators that nested project can be pushed through. `Generate` isn't one of them. So the general case doesn't work on this case.

Reply:
`Generate` is special for another reason. We can't prune an output from its child even just a nested field of the output is used in the top project list. The generator could use it.

Reply:
Shall we add `if SQLConf.get.nestedSchemaPruningEnabled`?

Reply:
`nestedSchemaPruningEnabled` is for pruning nested fields from logical relation. But this fix isn't due to the same cause. For the data sources can't be pruned nested fields, it is also useful to apply this fix.

Reply:
Out of curiosity, why don't we need another configuration then if `spark.sql.optimizer.nestedSchemaPruning.enabled` isn't part of this optimization, or fixing the doc of `spark.sql.optimizer.nestedSchemaPruning.enabled`?

Reply:
This looks a general fix no matter the setting of `nestedSchemaPruning` is. Do we need a config to disable this?

Reply:
My impression was that we need a configuration but I think you or @dongjoon-hyun have more context then me about nested pruning stuff. @cloud-fan, @dongjoon-hyun, @gatorsmile, can you make a call here if we need a config or not?

Reply:
`nestedSchemaPruning` is vague as it can point to nested pruning in scan, or general nested pruning in other operators. Currently the config affects scan nested pruning. I feel it is not tightly related to the fix here, because the fix isn't for scan.

If we are considering a config here, I think a different config or fix the doc of `nestedSchemaPruning` to explicitly indicate it also helps general nested pruning.

Reply:
Let me fix the doc of `nestedSchemaPruning` and apply this pruning when `nestedSchemaPruning` is enabled.